builder/llpcBuilderImplSubgroup.cpp

```diff
@@ -0,0 +1,1185 @@
+            return ConstantFP::getInfinity(pType, true);
+        case GroupOperation::And:
+            return ConstantInt::get(pType, UINT64_MAX, false);
+        case GroupOperation::Or:
+            return ConstantInt::get(pType, 0, false);
+        case GroupOperation::Xor:
+            return ConstantInt::get(pType, 0, false);
+    }
+}
+
+// =====================================================================================================================
+// Create the group operation arithmetic on x and y.
+Value* BuilderImplSubgroup::CreateGroupOperationArithmetic(
+    GroupOperation groupOperation, // The group operation to use for the reduction.
+    Value* const   pX,             // [in] The x value to perform the arithmetic on.
+    Value* const   pY)             // [in] The y value to perform the arithmetic on.
+{
+    switch (groupOperation)
+    {
+    default:
+        LLPC_NOT_IMPLEMENTED();
+        return nullptr;
+    case GroupOperation::IAdd:
+        return CreateAdd(pX, pY);
+    case GroupOperation::FAdd:
+        return CreateFAdd(pX, pY);
+    case GroupOperation::IMul:
+        return CreateMul(pX, pY);
+    case GroupOperation::FMul:
+        return CreateFMul(pX, pY);
+    case GroupOperation::SMin:
+        return CreateSelect(CreateICmpSLT(pX, pY), pX, pY);
+    case GroupOperation::UMin:
+        return CreateSelect(CreateICmpULT(pX, pY), pX, pY);
+    case GroupOperation::FMin:
+        return CreateMinNum(pX, pY);
+    case GroupOperation::SMax:
+        return CreateSelect(CreateICmpSGT(pX, pY), pX, pY);
+    case GroupOperation::UMax:
+        return CreateSelect(CreateICmpUGT(pX, pY), pX, pY);
+    case GroupOperation::FMax:
+        return CreateMaxNum(pX, pY);
+    case GroupOperation::And:
+        return CreateAnd(pX, pY);
+    case GroupOperation::Or:
+        return CreateOr(pX, pY);
+    case GroupOperation::Xor:
+        return CreateXor(pX, pY);
+    }
+}
+
+// =====================================================================================================================
+// Create an inline assembly call to cause a side effect (used to workaround mis-compiles with convergent).
+Value* BuilderImplSubgroup::CreateInlineAsmSideEffect(
+    Value* const pValue) // [in] The value to ensure doesn't move in control flow.
+{
+    Type* const pType = pValue->getType();
+
+    // Some vector types don't play nice with inline assembly, so for those we just do the inline assembly trick on the
+    // 0th element of the vector.
+    if (pType->isVectorTy())
+    {
+        Value* const pExtract0 = CreateExtractElement(pValue, static_cast<uint64_t>(0));
+        Value* const pSideEffect = CreateInlineAsmSideEffect(pExtract0);
+        return CreateInsertElement(pValue, pSideEffect, static_cast<uint64_t>(0));
+    }
+    else if (pType->getPrimitiveSizeInBits() < 32)
+    {
+        Type* const pVectorType = VectorType::get(pType, (pType->getPrimitiveSizeInBits() == 16) ? 2 : 4);
```

Comment:
I have a little confused, when will you create vec x 4?  pType->getPrimitiveSizeInBits() == 8?

Reply:
Yes!